File: hp9k_serial/phaseb_probe.py
#!/usr/bin/env python3
"""
Phase B step 1: validate bidirectional serial and capture the guest /etc/ttys over the (now working)
serial line. Drives phaseb_probe.lua: the guest cats /etc/ttys to /dev/tty0, then blocks reading a line
from /dev/tty0; this host sends "PING" and checks it echoes back as __RXGOT__[PING] -> RX confirmed.
"""
import sys, os, re
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from hp_ser_io import HpSerial

HERE = os.path.dirname(os.path.abspath(__file__))
LUA = os.path.join(HERE, "phaseb_probe.lua")


def main():
    hs = HpSerial(log_path=os.path.join(HERE, "phaseb.log"),
                  mame_log_path=os.path.join(HERE, "phaseb_mame.log")).listen()
    hs.launch_mame(lua=LUA, seconds=300, video="none")
    if not hs.accept():
        print("ACCEPT FAIL:", hs.connect_error); hs.close(); return 2
    print("[connected]")

    if not hs.expect("__TTYS_END__", timeout=180):
        print("!! never saw __TTYS_END__ (guest may not have reached the dump)")
    r = hs.expect("__RXREADY__", timeout=60)
    rx_ok = False
    if r:
        print("[guest is reading /dev/tty0 -> sending PING from host]")
        hs.send("PING")                                  # host -> guest, eol=\r
        rx_ok = hs.expect(r"__RXGOT__\[PING", timeout=40) is not None
    else:
        print("!! never saw __RXREADY__")

    hs.wait_mame(timeout=90)
    hs.drain_and_join()
    full = hs.text()

    print("\n" + "=" * 64)
    print("RX (host->guest): %s" % ("CONFIRMED" if rx_ok else "NOT CONFIRMED"))
    idx = full.find("__TTYS_END__")
    print("--- guest /etc/ttys (received over serial) ---")
    if idx > 0:
        # strip the connect banner + show the ttys body
        body = full[:idx]
        body = body.split("]\n", 1)[-1] if "]\n" in body[:60] else body
        print(body.strip()[-1600:])
    else:
        print("(not captured)")
    print("--- RX echo line ---")
    for line in full.splitlines():
        if "__RXGOT__" in line:
            print("  " + line)
    print("=" * 64)
    hs.close()
    return 0 if rx_ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: hp9k_serial/phaseb_probe.lua
-- Phase B step 1: TX now works, so use the serial line itself to (a) dump the guest /etc/ttys to the
-- host over /dev/tty0, and (b) round-trip test RX (host->guest): print __RXREADY__, block reading a line
-- from /dev/tty0, echo it back as __RXGOT__[...]. If the host's "PING" comes back, RX is confirmed and the
-- interrupt fix is fully bidirectional (prerequisite for a getty).
local phase, np = 0, 0
local function snap() manager.machine.video:snapshot() end
emu.register_periodic(function()
    local t = emu.time()
    local nk = manager.machine.natkeyboard
    if     phase==0 and t>=205 then nk:post("root\r"); phase=1; np=t+6
    elseif phase==1 and t>=np  then nk:post("\r"); phase=2; np=t+4            -- TERM = (default)
    elseif phase==2 and t>=np  then nk:post("/bin/sh\r"); phase=3; np=t+4     -- csh -> sh
    elseif phase==3 and t>=np  then nk:post("stty sane\r"); phase=4; np=t+4
    elseif phase==4 and t>=np  then
        nk:post("cat /etc/ttys > /dev/tty0; echo __TTYS_END__ > /dev/tty0\r"); phase=5; np=t+12
    elseif phase==5 and t>=np  then snap()
        nk:post("echo __RXREADY__ > /dev/tty0; read L < /dev/tty0; echo __RXGOT__[$L] > /dev/tty0\r"); phase=6; np=t+30
    elseif phase==6 and t>=np  then snap(); phase=7
    end
end)
